les/fetcher.go

```diff
@@ -435,72 +435,131 @@ func (f *lightFetcher) nextRequest() (*distReq, uint64) {
 	var rq *distReq
 	reqID := genReqID()
 	if f.syncing {
-		rq = &distReq{
-			getCost: func(dp distPeer) uint64 {
-				return 0
-			},
-			canSend: func(dp distPeer) bool {
-				p := dp.(*peer)
-				f.lock.Lock()
-				defer f.lock.Unlock()
-
-				fp := f.peers[p]
-				return fp != nil && fp.nodeByHash[bestHash] != nil
-			},
-			request: func(dp distPeer) func() {
-				go func() {
-					p := dp.(*peer)
-					p.Log().Debug("Synchronisation started")
-					f.pm.synchronise(p)
-					f.syncDone <- p
-				}()
-				return nil
-			},
-		}
+		rq = f.newFetcherDistReqForSync(bestHash)
 	} else {
-		rq = &distReq{
-			getCost: func(dp distPeer) uint64 {
-				p := dp.(*peer)
-				return p.GetRequestCost(GetBlockHeadersMsg, int(bestAmount))
-			},
-			canSend: func(dp distPeer) bool {
+		rq = f.newFetcherDistReq(bestHash, reqID, bestAmount)
+	}
+	return rq, reqID
+}
+
+// findBestValues retrieves the best values for LES or ULC mode.
+func (f *lightFetcher) findBestValues() (bestHash common.Hash, bestAmount uint64, bestTd *big.Int, bestSyncing bool) {
```

Comment:
The name "findBestValues" does not tell anything about what kind of values it finds, please rename. My suggestion would by something like "findBestRequest". Also, since this is one of the most important parts of the fetcher logic, now that it is a separate function it deserves a proper description too. This is what I suggest:

findBestRequest finds the best head to request that has been announced by but not yet requested from a known peer. It also returns the announced Td (which should be verified after fetching the head), the necessary amount to request and whether a downloader sync is necessary instead of a normal header request.